File: algorithms/sorting/radix_sort/radix_sort.cpp
#include <iostream>

using namespace std;

#define REP(i, n) for(int i = 0; i < n; i++)

int get_digit(int num, int pos)
{
  string c = to_string(num);
  if (pos >= c.length()) return 0;
  else {
    int new_pos = c.length() -1 + pos * -1;
    return (c[new_pos] - '0');
  }
}

void counting_sort(int *a, int size, int max, int pos)
{

  int *c = new int[max + 1];
  int *b = new int[size];

  REP(i, max + 1) c[i] = 0;
  REP(j, size) {
    int digit = get_digit(a[j], pos);
    ++c[digit];
  }
  
  for(int i = 0; i <= max; ++i)
    c[i] = c[i] + c[i - 1];

  for(int j = size - 1; j >= 0; --j) {
    int digit = get_digit(a[j], pos);
    b[ c[digit] - 1] = a[j];
    --c[digit];
  }

  REP(i, size) a[i] = b[i];
  delete [] c;
  delete [] b;
}

int get_max(int *a, int size)
{
  int largest = a[0];
  REP(i, size) {
    if (a[i] > largest)
      largest = a[i];
  }
  return largest;
}

void radix_sort(int *a, int size)
{
  int max = get_max(a, size);
  string c = to_string(max);

  REP(i, c.length()) {
    counting_sort(a, size, 9, i);
  }
}

int main()
{
  int a[] = { 1811, 76, 92, 43, 890, 567, 7, 289 };

  radix_sort(a, 8);

  REP(i, 8) cout << a[i] << "  ";
  cout << endl;

  return 0;
}

File: algorithms/sorting/radix_sort/radix_sort.rb
class Array

  public

    def radix_sort
      d = self.max.to_s.length
      array = self
      (0...d).each do |digit_pos|
        array = counting_sort(array, 9, digit_pos)
      end
      array
    end

    def counting_sort(array, max_element, digit_pos)
      c = []
      b = []

      (0..max_element).each { c.push(0) }

      (0...size).each do |j| 
        digit = get_digit(array[j], digit_pos)
        c[ digit ] += 1
      end

      (1..max_element).each { |i| c[i] = c[i] + c[i - 1] }

      (0...size).to_a.reverse.each do |j|
        digit = get_digit(array[j], digit_pos)
        b[ c[digit] - 1 ] = array[j]
        c[digit] -= 1
      end

      b
    end

    def get_digit(num, i)
      num_s = num.to_s
      return 0 if i >= num_s.length
      num_s[-1 + i*-1].to_i
    end
end

a = [1811, 76, 92, 43, 890, 567, 7, 289]

puts a.radix_sort.inspect
